gateway/ln-gateway/src/db.rs

```diff
@@ -501,13 +520,89 @@ async fn migrate_to_v4(mut ctx: MigrationContext<'_>) -> Result<(), anyhow::Erro
                 transaction_fee: PaymentFee::TRANSACTION_FEE_DEFAULT,
                 connector: Connector::default(),
             };
-            let new_key = FederationIdKey { id: fed_id.id };
+            let new_key = FederationConfigKey { id: fed_id.id };
             dbtx.insert_new_entry(&new_key, &new_fed_config).await;
         }
     }
     Ok(())
 }
 
+/// Introduced in v0.5, there is a db key clash between the `FederationConfig`
+/// record and the isolated databases used for each client. We must migrate the
+/// isolated databases to be behind the `ClientDatabase` prefix to allow the
+/// gateway to properly read the federation configs.
+async fn migrate_to_v5(mut ctx: MigrationContext<'_>) -> Result<(), anyhow::Error> {
+    let mut dbtx = ctx.dbtx();
+    migrate_federation_configs(&mut dbtx).await
+}
+
+async fn migrate_federation_configs(
+    dbtx: &mut DatabaseTransaction<'_>,
+) -> Result<(), anyhow::Error> {
+    // We need to migrate all isolated database entries to be behind the 0x10
+    // prefix. The problem is, if there is a `FederationId` that starts with
+    // 0x04, we cannot read the `FederationId` because the database will be confused
+    // between the isolated DB and the `FederationConfigKey` record. To solve this,
+    // we try and decode each key as a Federation ID and each value as a
+    // FederationConfig. If that is successful and the federation ID in the
+    // config matches the key, then we skip that record and migrate the rest of
+    // the entries.
+    let problem_entries = dbtx
```

Comment:
`s/problem_entries/raw_federation_configs`?

Reply:
They might not be federation configs. They could just be from the isolated database.

Reply:
There might also not be "problem entries", but just valid entries. Most of them, most of the time will be just valid configs.


The way I think about it - is we read through "raw federation configs" to find "valid ones".


I would actually do it other way around, even more defensively:

* go through all (probably) raw configs
* filter out invalid ones, and collect valid ones, deserializing them along the way
* then for all the valid deserialized ones, do the migration
* after all is done, do a normal prefixed "read all federation configs"
  * for every federation config, maybe even read some federation-prefixed key from the database to validate that it migrated right